include/net/cloud.h

```diff
@@ -0,0 +1,246 @@
+enum cloud_event_type {
+	CLOUD_EVT_CONNECTED,
+	CLOUD_EVT_DISCONNECTED,
+	CLOUD_EVT_READY,
+	CLOUD_EVT_ERROR,
+	CLOUD_EVT_DATA_SENT,
+	CLOUD_EVT_DATA_RECEIVED,
+	CLOUD_EVT_COUNT
+};
+
+/**@brief Quality of Service for message sent by a cloud backend. */
+enum cloud_qos {
+	CLOUD_QOS_AT_MOST_ONCE,
+	CLOUD_QOS_AT_LEAST_ONCE,
+	CLOUD_QOS_EXACTLY_ONCE,
+	CLOUD_QOS_COUNT
+};
+
+/**@brief Quality of Service for message sent by a cloud backend. */
+enum cloud_resource {
+	CLOUD_EP_TOPIC,
+	CLOUD_EP_URI,
+	CLOUD_EP_COUNT
+};
+
+/* Forward declaration of cloud backend type. */
+struct cloud_backend;
+
+/**@brief Cloud message type. */
+struct cloud_msg {
+	char *payload;
+	size_t len;
+	enum cloud_qos qos;
+	struct {
+		enum cloud_resource type;
+		char *str;
+		size_t len;
+	} resource;
+};
+
+/**@brief Cloud event type. */
+struct cloud_event {
+	enum cloud_event_type type;
+	union {
+		struct cloud_msg msg;
+		int err;
+	} data;
+};
+
+/**
+ * @brief Cloud event handler function type.
+ *
+ * @param[in] backend Pointer to cloud backend.
+ * @param[in] evt Pointer to cloud event.
+ */
+typedef void (*cloud_evt_handler_t)(const struct cloud_backend *const backend,
+				    const struct cloud_event *const evt);
+
+/**
+ * @brief Cloud backend API.
+ *
+ * All cloud backends must expose this API.
+ */
+struct cloud_api {
+	int (*init)(struct cloud_backend *const backend,
+		    cloud_evt_handler_t handler);
+	int (*uninit)(const struct cloud_backend *const backend);
+	int (*connect)(const struct cloud_backend *const backend);
+	int (*disconnect)(const struct cloud_backend *const backend);
+	int (*send)(const struct cloud_backend *const backend,
+			 const struct cloud_msg *const msg);
+	int (*ping)(const struct cloud_backend *const backend);
+	int (*input)(const struct cloud_backend *const backend);
+};
+
+/**@brief Structure for cloud backend configuration. */
+struct cloud_backend_config {
+	char *name;
+	cloud_evt_handler_t handler;
+	int socket;
+};
+
+/**@brief Structure for cloud backend. */
+struct cloud_backend {
+	struct cloud_api *api;
+	struct cloud_backend_config *config;
+};
+
+/**@brief API to initialize a cloud backend. Performs all necessary
+ *	  configuration of the backend to be able to connect its online
+ *	  counterpart.
+ *
+ * param[in] cloud Pointer to cloud backend structure.
+ * param[in] ctx Pointer to backend specific data.
+ */
+static inline int cloud_init(
+	struct cloud_backend *const backend,
+	cloud_evt_handler_t handler)
+{
+	if (backend == NULL) {
+		return -EINVAL;
+	}
+
+	if (handler == NULL) {
+		return -EINVAL;
+	}
+
+	return backend->api->init(backend, handler);
+}
+
+/**@brief API to uninitialize a cloud backend. Gracefully disconnects
+ *        remote endpoint and releases memory.
+ *
+ * param[in] cloud Pointer to cloud backend structure.
+ */
+static inline int cloud_uninit(const struct cloud_backend *const backend)
+{
+	__ASSERT_NO_MSG(backend != NULL);
```

Comment:
And don't you want to check if the API function is defined? I think you should return an error if the `uninit` function is NULL. Same for other functions below, in case the function is mandatory to implement.

Reply:
I can do `if`s instead and return error codes if `NULL` so that it works for `CONFIG_ASSERT=n` as well